### PR TITLE
Walletlink v2

### PR DESCRIPTION
**Pull request by hishbase**
Update iOS SDK to handle WalletLink v2 protocol

### CODE REVIEW

File: ios/Source/APIs/WalletLinkConnection.swift [hunk 1]
```diff
@@ -118,6 +161,22 @@ class WalletLinkConnection {
             .asVoid()
     }
 
+    /// Destroy session
+    ///
+    /// - Parameters:
+
```
Comment (AndrewGold) [inline ios/Source/APIs/WalletLinkConnection.swift]: Nit: remove newline here

File: ios/Source/APIs/WalletLinkWebSocket.swift [hunk 2]
```diff
@@ -198,6 +203,16 @@ final class WalletLinkWebSocket {
             }
 
             receivedServerRequest(request)
+        case .getSessionConfigOK, .sessionConfigUpdated:
+            guard
+                let sessionId = json["sessionId"] as? String,
+                let metadata = json["metadata"] as? [String: Any],
+                let destroyedValue = metadata[ClientMetadataKey.destroyed.rawValue] as? String,
+                destroyedValue == "1"
```
Comment (AndrewGold) [inline ios/Source/APIs/WalletLinkWebSocket.swift]: Nit: constant for this value?

File: ios/Source/APIs/WalletLinkConnection.swift (hunk 1 above)
Comment (hishbase) [reply ios/Source/APIs/WalletLinkConnection.swift]: lol fixed

File: ios/Source/APIs/WalletLinkWebSocket.swift (hunk 2 above)
Comment (hishbase) [reply ios/Source/APIs/WalletLinkWebSocket.swift]: fixed

File: ios/Source/APIs/WalletLinkConnection.swift [hunk 3]
```diff
@@ -51,12 +56,41 @@ class WalletLinkConnection {
...
+                else {
+                    return request
+                }
+
+                // for WalletLink v > 1, grab dapp details from EIP1102 request
+                self.linkRepository.saveSession(
+                    url: session.url,
+                    sessionId: session.id,
+                    secret: session.secret,
+                    version: session.version,
```
Comment (johnnycoinbase) [inline ios/Source/APIs/WalletLinkConnection.swift]: nit - pass in just the session object to the `init`?

File: ios/Source/APIs/WalletLinkConnection.swift [hunk 4]
```diff
@@ -51,12 +56,41 @@ class WalletLinkConnection {
...
+
+                let hostRequestId = request.hostRequestId
+                let foundSession = self.linkRepository.sessions
+                    .first { $0.url == hostRequestId.url && $0.id == hostRequestId.sessionId }
+
+                guard
+                    let session = foundSession,
+                    session.version != nil,
+                    case .dappPermission = request
+                else {
```
Comment (johnnycoinbase) [inline ios/Source/APIs/WalletLinkConnection.swift]: nit - one-liner?

File: ios/Source/Module/WalletLink.swift [hunk 5]
```diff
@@ -77,16 +78,19 @@ public class WalletLink: WalletLinkProtocol {
...
             }
     }
 
-    public func unlink(session: Session) {
-        linkRepository.delete(url: session.url, sessionId: session.id)
+    public func unlink(session: Session) -> Single<Void> {
+        guard let connection = connections[session.url] else { return .error(WalletLinkError.noConnectionFound) }
+
+        return connection.destroySession(for: session.id)
+            .map { _ in self.linkRepository.delete(url: session.url, sessionId: session.id) }
```
Comment (alexshoykhet) [inline ios/Source/Module/WalletLink.swift]: The delete logic in SessionDAO is incorrect.
```
func delete(url: URL, sessionId: String) {
        accessQueue.sync {
            let sessions = self.store.get(.sessions)?.items
                .filter { $0.id != sessionId && $0.url == url } ?? []

            self.store.set(.sessions, value: SessionList(items: sessions))
        }
    }
```
Should be
```
func delete(url: URL, sessionId: String) {
        accessQueue.sync {
            let sessions = self.store.get(.sessions)?.items
                .filter { $0.id != sessionId || $0.url != url } ?? []

            self.store.set(.sessions, value: SessionList(items: sessions))
        }
    }
```

File: ios/Source/APIs/WalletLinkWebSocket.swift [hunk 6]
```diff
@@ -182,7 +187,7 @@ final class WalletLinkWebSocket {
             let typeString = json["type"] as? String,
             let type = ServerMessageType(rawValue: typeString)
         else {
-            return print("Unknown WalletLink type \(incoming)")
+            return print("[walletlink] Unknown WalletLink type \(incoming)")
```
Comment (johnnycoinbase) [inline ios/Source/APIs/WalletLinkWebSocket.swift]: nit - remove print?

File: ios/Source/APIs/WalletLinkWebSocket.swift [hunk 7]
```diff
@@ -198,6 +203,16 @@ final class WalletLinkWebSocket {
...
             receivedServerRequest(request)
+        case .getSessionConfigOK, .sessionConfigUpdated:
+            guard
+                let sessionId = json["sessionId"] as? String,
+                let metadata = json["metadata"] as? [String: Any],
+                let destroyedValue = metadata[ClientMetadataKey.destroyed.rawValue] as? String,
+                destroyedValue == .one
+            else { return }
+
+            print("[walletlink] destroy session \(sessionId)")
```
Comment (johnnycoinbase) [inline ios/Source/APIs/WalletLinkWebSocket.swift]: nit - remove print?

File: ios/Source/Extensions/String+WalletLink.swift [hunk 8]
```diff
@@ -0,0 +1,8 @@
+// Copyright (c) 2018-2019 Coinbase, Inc. <https://coinbase.com/>
+// Licensed under the Apache License, version 2.0
+
+import Foundation
+
+extension String {
+    internal static let one = "1"
```
Comment (johnnycoinbase) [inline ios/Source/Extensions/String+WalletLink.swift]: nit - can this be named more specific like `destroyID` or something?

File: ios/Source/Module/WalletLink.swift [hunk 9]
```diff
@@ -59,12 +59,13 @@ public class WalletLink: WalletLinkProtocol {
     public func link(
```
Comment (johnnycoinbase) [inline ios/Source/Module/WalletLink.swift]: nit - docstring?

File: ios/Source/Module/WalletLink.swift [hunk 10]
```diff
@@ -77,16 +78,19 @@ public class WalletLink: WalletLinkProtocol {
...
             .map { _ in self.observeConnection(connection) }
             .catchError { err in
                 self.connections[url] = nil
                 throw err
             }
     }
 
-    public func unlink(session: Session) {
-        linkRepository.delete(url: session.url, sessionId: session.id)
+    public func unlink(session: Session) -> Single<Void> {
```
Comment (johnnycoinbase) [inline ios/Source/Module/WalletLink.swift]: nit - docstring?

File: ios/Source/DAOs/SessionDAO.swift [hunk 11]
```diff
@@ -46,12 +46,33 @@ final class SessionDAO {
...
-    func save(url: URL, sessionId: String, secret: String) {
+    ///     - version: WalletLink server version
+    ///     - dappName: DApp name that initiated the new wallet link connection
+    ///     - dappImageURL: DApp image that initiated the new wallet link connection
+    ///     - dappURL: DApp URL that initiated the wallet link connection
+    func save(
+        url: URL,
+        sessionId: String,
+        secret: String,
+        version: String?,
```
Comment (johnnycoinbase) [inline ios/Source/DAOs/SessionDAO.swift]: what does it mean for a version to be nil?

Comment (hishbase) [reply ios/Source/DAOs/SessionDAO.swift]: prior to WL 2.0 we didn't pass a version. So version = nil is a WL 1.0

File: ios/Source/Module/WalletLink.swift (hunk 5 above)
Comment (hishbase) [reply ios/Source/Module/WalletLink.swift]: nice catch. fixed

File: ios/Source/APIs/WalletLinkWebSocket.swift [hunk 12]
```diff
@@ -198,6 +203,16 @@ final class WalletLinkWebSocket {
             }
 
             receivedServerRequest(request)
+        case .getSessionConfigOK, .sessionConfigUpdated:
+            guard
+                let sessionId = json["sessionId"] as? String,
```
Comment (alexshoykhet) [inline ios/Source/APIs/WalletLinkWebSocket.swift]: What do you think about adding a DeleteRequestDTO object?  On android, I need to do a bit of extra processing on the metadata map to make sure it's legit.

File: ios/Source/DAOs/SessionDAO.swift [hunk 13]
```diff
@@ -46,12 +46,33 @@ final class SessionDAO {
...
+        dappName: String?,
+        dappImageURL: URL?,
+        dappURL: URL?
+    ) {
         accessQueue.sync {
-            var sessions = self.store.get(.sessions)?.items
-                .filter { $0.id != sessionId && $0.url == url } ?? []
+            var sessions = self.store.get(.sessions)?.items.filter { $0.id != sessionId && $0.url == url } ?? []
```
Comment (alexshoykhet) [inline ios/Source/DAOs/SessionDAO.swift]: This logic looks incorrect too actually.  
I think it should be 
```
var sessions = self.store.get(.sessions)?.items.filter { $0.id != sessionId && $0.url != url } ?? []
```

**Pull request merged**